drivers/at25xxx/at25xxx.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2019 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_at25xxx
+ * @{
+ *
+ * @file
+ * @brief       Driver for the AT25xxx family of SPI-EEPROMs.
+ *              This also includes M95xxx, 25AAxxx, 25LCxxx,
+ *              CAT25xxx & BR25Sxxx.
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "at25xxx.h"
+#include "at25xxx_constants.h"
+#include "at25xxx_params.h"
+#include "byteorder.h"
+
+#ifdef USEMODULE_XTIMER
+#include "xtimer.h"
+#define POLL_DELAY_US   (1000)
+#endif
+
+#ifndef min
+#define min(a, b) ((a) > (b) ? (b) : (a))
+#endif
+
+#define PAGE_SIZE   (dev->params.page_size)
+#define ADDR_LEN    (AT25XXX_PARAM_ADDR_LEN)
+#define ADDR_MSK    ((1UL << ADDR_LEN) - 1)
+
+#ifndef AT25XXXX_SET_BUF_SIZE
+/**
+ * @brief  Adjust to configure buffer size
+ */
+#define AT225XXXX_SET_BUF_SIZE      (64)
+#endif
+
+static inline int getbus(const at25xxx_t *dev)
+{
+    return spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0, dev->params.spi_clk);
+}
+
+static inline uint32_t _pos(uint8_t cmd, uint32_t pos)
+{
+    /* first byte is CMD, then addr with MSB first */
+    pos = htonl((pos & ADDR_MSK) | ((uint32_t)cmd << ADDR_LEN));
+    pos >>= 8 * sizeof(pos) - (ADDR_LEN + 8);
+    return pos;
+}
+
+static inline bool _write_in_progress(const at25xxx_t *dev)
+{
+    return spi_transfer_reg(dev->params.spi, dev->params.cs_pin, CMD_RDSR, 0) & SR_WIP;
+}
+
+static inline bool _write_enabled(const at25xxx_t *dev)
+{
+    return spi_transfer_reg(dev->params.spi, dev->params.cs_pin, CMD_RDSR, 0) & SR_WEL;
+}
+
+static inline void _wait_until_eeprom_ready(const at25xxx_t *dev)
+{
+    while (_write_in_progress(dev)) {
+#ifdef USEMODULE_XTIMER
+        spi_release(dev->params.spi);
+        xtimer_usleep(POLL_DELAY_US);
+        getbus(dev);
+#endif
+    }
+}
+
+static size_t _write_page(const at25xxx_t *dev, uint32_t pos, const void *data, size_t len)
+{
+    /* write no more than to the end of the current page to prevent wrap-around */
+    len = min(len, PAGE_SIZE - (pos & (PAGE_SIZE - 1)));
+    pos = _pos(CMD_WRITE, pos);
+
+    /* wait for previous write to finish - may take up to 5 ms */
+    _wait_until_eeprom_ready(dev);
+
+    /* set write enable and wait for status change */
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, false, CMD_WREN);
+    while (!_write_enabled(dev)) {}
```

Comment:
Is there a need to have a poll counter that makes the function fail, if `SR_WEL` will never be set?
I could imagine that case if the EEPROM is not hooked up correctly. In that case `while (!_write_enabled(dev)) {}`
would spin forever. I thought the same for `_wait_until_eeprom_ready()`, but I think if SPI is set up incorrectly,
any `spi_transfer_reg()` will return `0` (or 255?), so `SR_WIP` will never be set and the EEPROM is always asserted to be ready, so it would not spin forever here.

Reply:
I've added 10 retries (10ms) with xtimer.
Without xtimer we'd have to know the SPI frequency to set good values. so I just resorted to the CPU clock instead.
It's probably too long of a wait for most cases, but then again the timeout should not happen in the first place.

Reply:
Ah, that was for `_wait_until_eeprom_ready()` - I added the timeout here too.